frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A lottery pallet that uses participation in the network to purchase tickets.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+mod benchmarking;
+pub mod weights;
+
+use sp_std::prelude::*;
+use sp_runtime::{DispatchError, ModuleId};
+use sp_runtime::traits::{AccountIdConversion, Saturating, Zero};
+use frame_support::{Parameter, decl_module, decl_error, decl_event, decl_storage, ensure, RuntimeDebug};
+use frame_support::dispatch::{Dispatchable, DispatchResult, GetDispatchInfo};
+use frame_support::traits::{
+	Currency, ReservableCurrency, Get, EnsureOrigin, ExistenceRequirement::KeepAlive, Randomness,
+};
+use frame_support::weights::Weight;
+use frame_system::ensure_signed;
+use codec::{Encode, Decode};
+pub use weights::WeightInfo;
+
+type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// The module's config trait.
+pub trait Config: frame_system::Config {
+	/// The Lottery's module id
+	type ModuleId: Get<ModuleId>;
+
+	/// A dispatchable call.
+	type Call: Parameter + Dispatchable<Origin=Self::Origin> + GetDispatchInfo + From<frame_system::Call<Self>>;
+
+	/// The currency trait.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// Something that provides randomness in the runtime.
+	type Randomness: Randomness<Self::Hash>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+
+	/// The manager origin.
+	type ManagerOrigin: EnsureOrigin<Self::Origin>;
+
+	/// The max number of calls available in a single lottery.
+	type MaxCalls: Get<usize>;
+
+	/// Weight information for extrinsics in this pallet.
+	type WeightInfo: WeightInfo;
+}
+
+type Index = u32;
```

Comment:
This maybe confused with https://github.com/paritytech/substrate/blob/dcf6a3ae9587bdf27a58eeff4bad8628c4763b38/frame/system/src/lib.rs#L180